### PR TITLE
Slightly simplify LastCommitCache

### PR DESCRIPTION
**Pull request by zeripath**
The LastCommitCache code is a little complex and there is unnecessary
duplication between the gogit and nogogit variants.

This PR adds the LastCommitCache as a field to the git.Repository and
pre-creates it in the ReferencesGit helpers etc. There has been some
simplification and unification of the variant code.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: modules/git/last_commit_cache.go [hunk 1]
```diff
@@ -19,16 +20,95 @@ type Cache interface {
...
 func (c *LastCommitCache) Put(ref, entryPath, commitID string) error {
 	if c == nil || c.cache == nil {
 		return nil
 	}
 	log.Debug("LastCommitCache save: [%s:%s:%s]", ref, entryPath, commitID)
-	return c.cache.Put(c.getCacheKey(c.repoPath, ref, entryPath), commitID, c.ttl())
+	return c.cache.Put(getCacheKey(c.repoPath, ref, entryPath), commitID, c.ttl())
+}
+
+// Get get the last commit information by commit id and entry path
```
Comment (zeripath) [inline modules/git/last_commit_cache.go]:
```suggestion
// Get gets the last commit information by commit id and entry path
```

File: modules/git/last_commit_cache.go [hunk 2]
```diff
@@ -19,16 +20,95 @@ type Cache interface {
...
+		return nil, err
+	}
+	if c.commitCache == nil {
+		c.commitCache = make(map[string]*Commit)
+	}
+	c.commitCache[commitID] = commit
+	return commit, nil
+}
+
+func (c *LastCommitCache) GetCommitByPath(commitID, entryPath string) (*Commit, error) {
```
Comment (zeripath) [inline modules/git/last_commit_cache.go:94]:
```suggestion
// GetCommitByPath gets the last commit for the entry in the provided commit
func (c *LastCommitCache) GetCommitByPath(commitID, entryPath string) (*Commit, error) {
```

**Review by silverwind — APPROVED** [review]

File: routers/api/v1/utils/git.go [hunk 3]
```diff
@@ -27,10 +30,26 @@ func ResolveRefOrSha(ctx *context.APIContext, ref string) string {
...
+	if ctx.Repo.GitRepo != nil && ctx.Repo.GitRepo.LastCommitCache == nil {
+		commitsCount, err := cache.GetInt64(ctx.Repo.Repository.GetCommitsCountCacheKey(ref, true), func() (int64, error) {
+			commit, err := ctx.Repo.GitRepo.GetCommit(sha)
+			if err != nil {
+				return 0, err
+			}
+			return commit.CommitsCount()
+		})
+		if err != nil {
+			log.Error("Unable to get commits count for %s in %s. Error: %v", sha, ctx.Repo.Repository.FullName(), err)
```
Comment (Gusted) [inline routers/api/v1/utils/git.go:47]: Should it return here? No need to create a cache if it's for some reason erroring.

Comment (zeripath) [reply routers/api/v1/utils/git.go:47]: Honestly it depends on whether there is any way a commitscount could return an error but not mean that the repo is hosed.

I don't know. In any case the newcommitcache function will check the count returned. If the count is zero and the minimum count is zero then even if the repo is hosed it's still only a shim.

The lastcommitcache struct here isn't actually anything more than a lightweight shim for calling stuff. Handling and dealing with a broken repo can be deferred to all of the other code that is already written for dealing with that.

File: web_src/less/shared/issuelist.less [hunk 4]
```diff
@@ -101,7 +101,7 @@
         padding-left: 5px;
       }
 
-      a.milestone {
+      a.milestone, a.project {
```
Comment (zeripath) [inline web_src/less/shared/issuelist.less]:
```suggestion
      a.milestone {
```

**Review by Gusted — APPROVED** [review]

**Comment by zeripath** [comment]
make lgtm work

**Pull request merged**